docs/specs/workflow-customization.md

```diff
@@ -0,0 +1,1486 @@
+    - authentication: primary_email_code
+      target_step: setup_email
+  - type: verify
+    target_step: setup_email
+  - type: identify
+    id: setup_phone
+    one_of:
+    - identification: phone
+  - type: authenticate
+    one_of:
+    - authentication: primary_sms_code
+      target_step: setup_phone
+  - type: verify
+    target_step: setup_phone
+  - type: authenticate
+    one_of:
+    - authentication: primary_password
+
+login_flows:
+- id: default_login_flow
+  steps:
+  - id: identify
+    type: identify
+    one_of:
+    - identification: phone
+    - identification: email
+  - type: authenticate
+    if: steps.identify.identification == "phone"
+    one_of:
+    - authentication: primary_sms_code
+    - authentication: primary_password
+  - type: authenticate
+    if: steps.identify.identification == "email"
+    one_of:
+    - authentication: primary_email_code
+    - authentication: primary_sms_code
+    - authentication: primary_password
+
+signup_login_flows:
+- id: default_signup_login_flow
+  steps:
+  - id: step
+    type: identify
+    one_of:
+    - identification: phone
+      login_flow: default_login_flow
+      signup_flow: phone_first
+    - identification: email
+      login_flow: default_login_flow
+      signup_flow: default_signup_flow
+```
+
+#### Use case example 3: Google
+
+```yaml
+identification_methods:
+- id: email
+  type: "login_id"
+  login_id:
+    type: "email"
+
+authentication_methods:
+- id: primary_password
+  kind: primary
+  type: password
+- id: secondary_sms_code
+  kind: secondary
+  type: oob_otp_sms
+  phone_otp_mode: "sms"
+- id: secondary_totp
+  kind: secondary
+  type: totp
+
+signup_flows:
+- id: default_signup_flow
+  steps:
+  - type: identify
+    one_of:
+    - identification: email
+  - type: authenticate
+    one_of:
+    - authentication: primary_password
+
+login_flows:
+- id: default_login_flow
+  steps:
+  - type: identify
+    one_of:
+    - identification: email
+  - type: authenticate
+    one_of:
+    - authentication: primary_password
+  - type: authenticate
+    one_of:
+    - authentication: secondary_totp
+    - authentication: secondary_sms_code
+```
+
+#### Use case example 4: The Club
+
+```yaml
+identification_methods:
+- id: email
+  type: "login_id"
+  login_id:
+    type: "email"
+- id: phone
+  type: "login_id"
+  login_id:
+    type: "phone"
+- id: username
+  type: "username"
+  login_id:
+    type: "username"
+
+authentication_methods:
+- id: primary_password
+  kind: primary
+  type: password
+- id: primary_sms_code
+  kind: primary
+  type: oob_otp_sms
+  phone_otp_mode: "sms"
+
+# signup_flows is omitted here because the exact signup flow is unknown.
+
+login_flows:
+- id: default_login_flow
+  steps:
+  - type: identify
+    one_of:
+    - identification: email
+    - identification: phone
+    - identification: username
+  - type: authenticate
+    one_of:
+    - authentication: primary_password
+    - authentication: primary_sms_code
```

Comment:
should we allow `target` of `oob_opt_sms` (and email counterpart) specify not just to a step, but also to a login ID or a custom attributes?

But I think that's very edge cases... most case would works by allow target = "*"

Reply:
I chose step because we do not know which login ID would be selected by the end-user. What would happen if the config reference a login ID that was not selected?

And we do not support authenticating with custom attributes. We treat all custom attributes opaque.

Reply:
I know step is needed for reference, I meant it seems we might need more than that...

E.g. would there be use cases where we want to specific only these list of phone no. could be used for `secondary_oob_otp_sms`? If yes I don't see how we can do it with current configurations.

Reply:
If I understand you correctly, you are referring to a feature like this.

```yaml
authentication_methods:
- id: phone1
  kind: secondary
  type: oob_otp_sms
  labels: ["some_special_label"]
```

And then authenticator created with reference to `phone1`WILL have labels `["some_special_label"]` attached to them and persisted in the database. Later on we can reference to `phone1` to select those authenticators.

We do not support attaching any labels to identity / authenticator at the moment. And this feature does not work with settings page. We do not allow customized workflow for the actions in the settings page.

I think this is out of the scope of this PR.